algorithms/HPdutility/R/hpdsample.R

```diff
@@ -0,0 +1,303 @@
+      }
+
+      # Copy data items into sampled data
+      si[startRow:endRow,] <- dk[idx,,drop = F]
+
+      update(si)
+    }, progress = trace, scheduler = 1)
+  }
+  colnames(sdata) <- colnames(data)
+  sdata 
+}
+
+.hpdsampledual <- function(data1, data2, nSamplePartitions, samplingRatio, trace) {
+  # Function to sample two darrays/dframes in conjunction
+  
+  if (!(is.darray(data1) || is.dframe(data1))) 
+    stop("'data1' must be a darray or dframe")
+
+  if (!(is.darray(data2) || is.dframe(data2))) 
+    stop("'data2' must be a darray or dframe")
+
+  if (nrow(data1) != nrow(data2)) 
+    stop("'data1' and 'data2' must be the same length")
+
+  if (npartitions(data1) != npartitions(data2)) 
+    stop("'data1' and 'data2' must have the same number of partitions")
+
+  ndcol1    <- ncol(data1)
+  ndcol2    <- ncol(data2)
+  # Contains the amount of data to be sampled from each data partition into the
+  # output partitions
+  psizes    <- ceiling(partitionsize(data1)[,1] * samplingRatio)
+  finalnspp <- sum(psizes) # Number of samples per partition 
+
+  d1dim     <- c(finalnspp * nSamplePartitions, ndcol1)
+  d1blocks  <- c(finalnspp, ndcol1)
+  d2dim     <- c(finalnspp * nSamplePartitions, ndcol2)
+  d2blocks  <- c(finalnspp, ndcol2)
+  # dobject to contain the sampled data from data1
+  sdata1 <- if (is.darray(data1))      darray(dim = d1dim, blocks = d1blocks) 
+            else if (is.dframe(data1)) dframe(dim = d1dim, blocks = d1blocks) 
+  # dobject to contain the sampled data from data2 
+  sdata2 <- if (is.darray(data2))      darray(dim = d2dim, blocks = d2blocks)
+            else if (is.dframe(data2)) dframe(dim = d2dim, blocks = d2blocks)
+
+  # We need metadata about which of the columns are factors so that we can
+  # keep track of factors locally within each partition: 
+  d1Levels <- if(is.dframe(data1)) .getGlobalLevels(data1)
+  d2Levels <- if(is.dframe(data2)) .getGlobalLevels(data2)
+
+  for (k in 1:npartitions(data1)) { # For each partition dk of the input data
+    # Iterate over the output partitions si in parallel and copy some portion of
+    # the data from partition dk to sample partition si. This copied data will
+    # be appended to the data copied from previous partitions
+    foreach(i, 1:nSamplePartitions, function(si1 = splits(sdata1, i),
+                                      si2 = splits(sdata2, i),
+                                      dk1 = splits(data1, k),
+                                      dk2 = splits(data2, k),
+                                      psizes = psizes,
+                                      k = k, 
+                                      d1Levels = d1Levels,
+                                      d2Levels = d2Levels,
+                                      .coerceFactorsToGlobalLevels =
+                                        .coerceFactorsToGlobalLevels,
+                                      .copyClassStructure = .copyClassStructure) {
+      if (nrow(dk1) != nrow(dk2)) 
```

Comment:
It is fine to keep this check. Nevertheless, if you add the check that I suggested at the beginning for comparing partition sizes, this condition will never become true.